inst/include/dust/dust.hpp

```diff
@@ -639,6 +808,201 @@ class Dust {
+// order to make the device copy possible. There's nothing really
+// complicated in these in practice.
+template <typename T>
+void run_particles(size_t step_start, size_t step_end, size_t n_particles,
+                   size_t n_state, size_t n_int, size_t n_real, size_t n_pars,
+                   typename T::real_t * state,
+                   std::vector<dust::shared_ptr<T>> shared,
+                   uint64_t * rng_state) {
+  typedef typename T::real_t real_t;
+  const size_t n_particles_each = n_particles / n_pars;
+
+  // Allocation for our internal storage. We won't ever move this to
+  // or from the device so we just pass the size.
+  std::vector<real_t> state_next_v(n_particles * n_state);
+  std::vector<int> internal_int_v(n_particles * n_int);
+  std::vector<real_t> internal_real_v(n_particles * n_real);
+  real_t * state_next = state_next_v.data();
+  int * internal_int = internal_int_v.data();
+  real_t * internal_real = internal_real_v.data();
+  real_t * ret = state; // for ease of bookkeeping on return
+
+  // omp here
+  for (size_t i = 0; i < n_particles; ++i) {
+    dust::interleaved<real_t> p_state(state, i, n_particles);
+    dust::interleaved<real_t> p_state_next(state_next, i, n_particles);
+    dust::interleaved<int> p_internal_int(internal_int, i, n_particles);
+    dust::interleaved<real_t> p_internal_real(internal_real, i, n_particles);
+    dust::interleaved<uint64_t> p_rng(rng_state, i, n_particles);
+    // TODO: this needs work before moving to the device, but it might
+    // not be that bad in practice.
+    dust::shared_ptr<T> p_shared = shared[i / n_particles_each];
+
+    dust::rng_state_t<real_t> rng_block = dust::get_rng_state<real_t>(p_rng);
+    for (size_t step = step_start; step < step_end; ++step) {
+      update_device<T>(step,
+                       p_state,
+                       p_internal_int,
+                       p_internal_real,
+                       p_shared,
+                       rng_block,
+                       p_state_next);
+      std::swap(p_state, p_state_next);
+      // dust::interleaved<real_t> tmp = p_state;
+      // p_state = p_state_next;
+      // p_state_next = tmp;
+    }
+    dust::put_rng_state(rng_block, p_rng);
+
+    // In the inner loop, the swap will keep the locally scoped
+    // interleaved variables updated, but the interleaved variables
+    // passed in have not yet been updated.  If an even number of
+    // steps have been run state will have been swapped back into the
+    // original place, but an on odd number of steps the passed
+    // variables need to be swapped.
+  }
+
+  // This is going to require a bit of work to get exactly right as
+  // sometimes we hold the correct data
+  if ((step_end - step_start) % 2 == 1) {
+    // this is easily done without stl with memcpy:
+    // memcpy(state, state_next, sizeof(real_t) * n_state);
+    std::copy_n(state_next, n_state * n_particles, state);
```

Comment:
I didn't think that you could do either of these from a kernel, which reminds me why I did the pointer swap here. 

Checking, there is `cudaMemcpyAsync` and `cudaMemsetAsync` which can be called in a kernel. However we might want to be careful with this – is calling them from one thread efficient, or would it be better to call these from the host?

Reply:
this one you might have to explain to me as I am a bit in the dark about how these memory allocations work in the CUDA model

Reply:
Fixed in ad21b25